src/OrleansEventSourcing/EventStorage/LogViewAdaptor.cs

```diff
@@ -0,0 +1,360 @@
+                });
+
+            return success ? updates.Length : 0;
+        }
+
+
+
+        /// <summary>
+        /// Describes a connection issue that occurred when loading events from the event store.
+        /// </summary>
+        [Serializable]
+        public class LoadEventsFailed : PrimaryOperationFailed
+        {
+            /// <inheritdoc/>
+            public override string ToString()
+            {
+                return $"load stream failed: caught {Exception.GetType().Name}: {Exception.Message}";
+            }
+        }
+
+
+        /// <summary>
+        /// Describes a connection issue that occurred when appending events to the event store.
+        /// </summary>
+        [Serializable]
+        public class AppendEventsFailed : PrimaryOperationFailed
+        {
+            /// <summary>
+            /// Whether this failure was caused by an optimistic-concurrency conflict
+            /// </summary>
+            public bool IsOptimisticConcurrencyConflict { get; set; }
+
+            /// <inheritdoc/>
+            public override string ToString()
+            {
+                return $"append to stream failed: caught {Exception.GetType().Name}: {Exception.Message}";
+            }
+        }
+
+
+        /// <summary>
+        /// A notification message that is sent to remote instances of this grain after the primary has been
+        /// updated, to let them know the latest version. Contains all the updates that were applied.
+        /// </summary>
+        [Serializable]
+        private class UpdateNotificationMessage : INotificationMessage
+        {
+            /// <inheritdoc/>
+            public int Version { get; set; }
+
+            /// <summary> The list of updates that were applied. </summary>
+            public List<KeyValuePair<Guid,TEvent>> Updates { get; set; }
+
+            /// <summary>
+            /// A representation of this notification message suitable for tracing.
+            /// </summary>
+            public override string ToString()
+            {
+                return string.Format("v{0} ({1} updates, last={2})", Version, Updates.Count, Updates.Last().Key);
+            }
+        }
+
+        private SortedList<long, UpdateNotificationMessage> notifications = new SortedList<long, UpdateNotificationMessage>();
+
+        /// <inheritdoc/>
+        protected override void OnNotificationReceived(INotificationMessage payload)
+        {
+            var um = payload as UpdateNotificationMessage;
+            if (um != null)
+                notifications.Add(um.Version - um.Updates.Count, um);
+            else
+                base.OnNotificationReceived(payload);
+        }
+
+        /// <inheritdoc/>
+        protected override void ProcessNotifications()
+        {
+
+            // discard notifications that are behind our already confirmed state
+            while (notifications.Count > 0 && notifications.ElementAt(0).Key < ConfirmedVersionInternal)
```

Comment:
the loop continuously gets the first element and then removes it so I cannot reuse across iterations. I could probably optimize this more but I am not sure it is important.